DistributedOptimizer: wait for all workers to finish _LocalOptimizer constructor

aazzolini opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30062 DistributedOptimizer: wait for all workers to finish _LocalOptimizer constructor**

This allows to catch exceptions during optimizer creation.

Differential Revision: [D18586108](https://our.internmc.facebook.com/intern/diff/D18586108/)

aazzolini committed:
DistributedOptimizer: wait for all workers to finish _LocalOptimizer constructor

This allows to catch exceptions during optimizer creation.

Differential Revision: [D18586108](https://our.internmc.facebook.com/intern/diff/D18586108/)

[ghstack-poisoned]

aazzolini committed:
Update on "DistributedOptimizer: wait for all workers to finish _LocalOptimizer constructor"


This allows to catch exceptions during optimizer creation.

Differential Revision: [D18586108](https://our.internmc.facebook.com/intern/diff/D18586108/)

[ghstack-poisoned]

mrshenli commented on torch/distributed/optim/optimizer.py:
```diff
@@ -68,16 +76,17 @@ def __init__(self, optimizer_class, params_rref, *args, **kwargs):
+        remote_optim_futs = []
         for worker, param_rrefs in per_worker_params_rref.items():
-            remote_optim_rref = rpc.remote(
+            remote_optim_rref = rpc.rpc_async(
```
this is actually a future, shall we rename this var?

mrshenli reviewed (approved):
LGTM

aazzolini committed:
Update on "DistributedOptimizer: wait for all workers to finish _LocalOptimizer constructor"


This allows to catch exceptions during optimizer creation.

Differential Revision: [D18586108](https://our.internmc.facebook.com/intern/diff/D18586108/)

[ghstack-poisoned]
